File: modules/core/src/main/java/turbolift/internals/engine/concurrent/atomic/AtomicRefVH.java
package turbolift.internals.engine.concurrent.atomic;
import java.lang.invoke.VarHandle;
import java.lang.invoke.MethodHandles;


public class AtomicRefVH<T> {
  private T valueVH;
  private static final VarHandle nameVH;

  static {
    try {
      nameVH = MethodHandles.lookup().findVarHandle(AtomicRefVH.class, "valueVH", Object.class);
    } catch (ReflectiveOperationException e) {
      throw new ExceptionInInitializerError(e);
    }
  }

  public AtomicRefVH(T a) {
    valueVH = a;
  }

  protected T getVH() {
    return (T) nameVH.get(this);
  }

  protected void setVH(T a) {
    nameVH.set(this, a);
  }

  protected T gasVH(T a) {
    return (T) nameVH.getAndSet(this, a);
  }

  protected boolean casVH(T a, T b) {
    return nameVH.compareAndSet(this, a, b);
  }

  protected T caxVH(T a, T b) {
    return (T) nameVH.compareAndExchange(this, a, b);
  }
}

File: modules/core/src/main/scala/turbolift/internals/engine/concurrent/atomic/AtomicRefVH.scala
package turbolift.internals.engine.concurrent.atomic
import java.lang.invoke.{VarHandle, MethodHandles}


private[turbolift] class AtomicRefVH[T](protected var unsafeValue: T):
  protected def getVH: T = AtomicRefVH.theVH.get(this)
  
  protected def setVH(a: T): Unit = AtomicRefVH.theVH.set(this, a)
  
  protected def gasVH(a: T): T = AtomicRefVH.theVH.getAndSet(this, a)

  protected def casVH(a: T, b: T): Boolean = AtomicRefVH.theVH.compareAndSet(this, a, b)

  protected def caxVH(a: T, b: T): T = AtomicRefVH.theVH.compareAndExchange(this, a, b)


private[turbolift] object AtomicRefVH:
  private val theVH: VarHandle = MethodHandles
    .privateLookupIn(classOf[AtomicRefVH[?]], MethodHandles.lookup()).nn
    .findVarHandle(classOf[AtomicRefVH[?]], "unsafeValue", classOf[Object]).nn
